pkg/openfeature/event_handler.go

```diff
@@ -0,0 +1,243 @@
+package openfeature
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	"golang.org/x/sync/errgroup"
+	"sync"
+	"time"
+)
+
+// event handler is registry to connect API and Client event handlers to Providers
+
+// handlerExecutionTime defines the maximum time event handler will wait for its handlers to complete
+const handlerExecutionTime = 500 * time.Millisecond
+
+type eventHandler struct {
+	providerShutdownHook map[string]chan interface{}
+	apiRegistry          map[EventType][]EventCallBack
+	clientRegistry       map[string]clientHolder
+	logger               logr.Logger
+	mu                   sync.Mutex
+}
+
+func newEventHandler(logger logr.Logger) eventHandler {
+	return eventHandler{
+		providerShutdownHook: map[string]chan interface{}{},
+		apiRegistry:          map[EventType][]EventCallBack{},
+		clientRegistry:       map[string]clientHolder{},
+		logger:               logger,
+	}
+}
+
+// clientHolder is a helper struct to hold client specific callbacks
+type clientHolder struct {
+	name   string
+	holder map[EventType][]EventCallBack
+}
+
+func newClientHolder(client string) clientHolder {
+	return clientHolder{
+		name:   client,
+		holder: map[EventType][]EventCallBack{},
+	}
+}
+
+func (e *eventHandler) updateLogger(l logr.Logger) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.logger = l
+}
+
+func (e *eventHandler) registerApiHandler(t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if e.apiRegistry[t] == nil {
+		e.apiRegistry[t] = []EventCallBack{c}
+	} else {
+		e.apiRegistry[t] = append(e.apiRegistry[t], c)
+	}
+}
+
+func (e *eventHandler) removeApiHandler(t EventType, c EventCallBack) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	entrySlice, ok := e.apiRegistry[t]
```

Comment:
can you also add a test for removing a handler that was not registered before?

Reply:
good suggestion, done :)

Reply:
Tests added to see if this panic